samples: Bluetooth: Add BLE Periodic Advertising sample

cvinayak opened the pull request:
Add an application that initializes Bluetooth Subsystem,
creates an extended advertising set, and starts Periodic
Advertising functionality.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: Add BLE Periodic Advertising sample

Add an application that initializes Bluetooth Subsystem,
creates an extended advertising set, and starts Periodic
Advertising functionality.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: Add BLE Periodic Advertising Sync sample

Add an application that initializes Bluetooth Subsystem,
starts scanning for Periodic Advertising, establishes
Periodic Advertising Sync and receives Periodic
Advertising Reports.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Add helper defines for Extended Advertising parameters

Add helper defines for Extended Advertising parameters that
can be used while creating Extended Advertising sets.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on include/bluetooth/bluetooth.h:
```diff
@@ -638,6 +638,23 @@ struct bt_le_per_adv_param {
 						 NULL)
 
+/** Non-connectable extended advertising with private address */
+#define BT_LE_EXT_ADV_NCONN BT_LE_ADV_PARAM(BT_LE_ADV_OPT_EXT_ADV, \
+					    BT_GAP_ADV_FAST_INT_MIN_2, \
+					    BT_GAP_ADV_FAST_INT_MAX_2, NULL)
+
+/** Non-connectable extended advertising with @ref BT_LE_ADV_OPT_USE_NAME */
+#define BT_LE_EXT_ADV_NCONN_NAME BT_LE_ADV_PARAM(BT_LE_ADV_OPT_EXT_ADV | \
+						 BT_LE_ADV_OPT_USE_NAME, \
+						 BT_GAP_ADV_FAST_INT_MIN_2, \
+						 BT_GAP_ADV_FAST_INT_MAX_2, \
+						 NULL)
+
+/** Non-connectable extended advertising with @ref BT_LE_ADV_OPT_USE_IDENTITY */
+#define BT_LE_EXT_ADV_NCONN_IDENTITY \
+	BT_LE_ADV_PARAM(BT_LE_ADV_OPT_EXT_ADV | BT_LE_ADV_OPT_USE_IDENTITY, \
+	BT_GAP_ADV_FAST_INT_MIN_2, BT_GAP_ADV_FAST_INT_MAX_2, NULL)
```
I was actually considering to add these as well, as I was missing them as well :D Glad to see them added

Thalley commented on samples/bluetooth/per_adv/src/main.c:
```diff
@@ -0,0 +1,55 @@
+
+void main(void)
+{
+	struct bt_le_ext_adv_start_param adv_start_param = {
+		.timeout = 0,
+		.num_events = 0,
+	};
```
Seeing as you added the other macros for extended adv, perhaps similar macros for the ext_adv_start_param could be added? I suppose most adv will just advertise without num_events or timeout so it would be generally useful. Can be added in another PR.

cvinayak replied:
I thought of adding, but as the Zephyr controller does not support connectable over extended advertising yet, I dont have/know any means to test it manually or in CI, hence not included in this PR. Yes, can be added in another PR.

Thalley replied:
I'm sure I understand: What does the `bt_le_ext_adv_start_param` have to do with connectable advertising? I guess you could test such a macro with the sample added in this PR, couldn't you?

cvinayak replied:
Correction, my mistake, I was thinking of `BT_LE_ADV_CONN`et al. 

Yes, `bt_le_ext_adv_start_param` can have a define. Will give a try.

Thalley reviewed (approved):
LGTM

Thalley reviewed (approved):
LGTM

rugeGerritsen commented on include/bluetooth/bluetooth.h:
```diff
@@ -637,6 +637,48 @@ struct bt_le_per_adv_param {
+#define BT_LE_EXT_ADV_NCONN BT_LE_ADV_PARAM(BT_LE_ADV_OPT_EXT_ADV, \
+					    BT_GAP_ADV_FAST_INT_MIN_2, \
+					    BT_GAP_ADV_FAST_INT_MAX_2, NULL)
+
```
Now that you are already adding these new APIs, maybe you should also add `BT_LE_EXT_ADV_NCONN_CODED`?
Maybe also apply the same pattern to `BT_LE_EXT_ADV_NCONN_NAME` and `BT_LE_EXT_ADV_NCONN_IDENTITY`

cvinayak replied:
I now have included the sample to sync with the periodic advertising, will add the defines for coded PHY and push the changes once I am done testing.

rugeGerritsen reviewed (commented):
Maybe we should split this review into two parts? The ext. adv. API updates are not strictly related to the sample itself. 

Should the sample documentation include information on how to test it? Now the only user feedback is "Starting Periodic Advertising Demo". Are there phones or other devices that can intercept these periodic advertising packets?
I would assume the sample documentation needs some more information on how it can be used or validated by the user.

cvinayak commented:
> Maybe we should split this review into two parts? The ext. adv. API updates are not strictly related to the sample itself.

Keeping the defines and samples in separate PR does not make sense, as PR just for defines not used anywhere would need a sample anyway for testing.

> Should the sample documentation include information on how to test it? Now the only user feedback is "Starting Periodic Advertising Demo". Are there phones or other devices that can intercept these periodic advertising packets?
I would assume the sample documentation needs some more information on how it can be used or validated by the user.

Yes, will update the documentation as I send PRs for sync sample. This is still a draft PR and commits to add a sample to generate periodic advertising report will make it big (still waiting for other open PRs to be merged). Keeping all the code in single PR will make it more understandable, but it is not desired to have big PRs.

cvinayak committed:
Bluetooth: controller: Fix maximum PDU len configuration

Fix maximum PDU len configuration when advertising and/or
scanning on primary channels.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Use memmove to copy overlapping memory

Use memmove to copy overlapping memory when populating the
Common Extended Advertising Header Format in the advertising
PDU.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
> the sample documentation needs some more information on how it can be used or validated by the user.

Yes, will include more information and update the PR soon.

cvinayak committed:
samples: Bluetooth: Add LED indications to PA Sync sample

Add LED blinking while scanning for Periodic Advertising,
and LED ON when Periodic Advertising Sync is established.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix Extended Scanning on Coded PHY

Fix Extended Scanning on Coded PHY, auxiliary PDUs on Coded
PHY where not received due to redundant conditional code
only applicable in connection role being copy-pasted during
the development of Extended Scanning feature.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: How to test periodic advertising

Added documentation in per_adv and per_sync samples on how
to test Periodic Advertising and Periodic Advertising
Synchronization.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on samples/bluetooth/per_sync/src/main.c:
```diff
@@ -0,0 +1,289 @@
+		if (err) {
+			printk("failed (err %d)\n", err);
+
+			goto per_adv_wait;
```
Since this is within a do { ... } while, could you not remove the goto and the label and use `continue` instead?

cvinayak replied:
Replaced code to use `continue`

Thalley commented on samples/bluetooth/per_sync/src/main.c:
```diff
@@ -0,0 +1,289 @@
+		}
+		printk("Found periodic advertising.\n");
+
+		printk("Creating Periodic Advertising Sync...");
+		bt_addr_le_copy(&sync_create_param.addr, &per_addr);
+		sync_create_param.options = 0;
+		sync_create_param.sid = per_sid;
+		sync_create_param.skip = 0;
+		sync_create_param.timeout = 0xa;
+		err = bt_le_per_adv_sync_create(&sync_create_param, &sync);
+		if (err) {
+			printk("failed (err %d)\n", err);
+			return;
+		}
+		printk("success.\n");
```
Did you consider putting this part in the scan callback? So rather than using a semaphore and a global variable, you just create the sync in the callback instead.

cvinayak replied:
Calling APIs in callback increases internal rx thread call stack usage. Keeping the API calls together improves readability in terms of demonstrating API usage in this sample.

rugeGerritsen commented on samples/bluetooth/per_sync/src/main.c:
```diff
@@ -0,0 +1,287 @@
+	bt_data_parse(buf, data_cb, name);
+
+	bt_addr_le_to_str(info->addr, le_addr, sizeof(le_addr));
+	printk("[DEVICE]: %s, AD evt type %u, Tx Pwr: %i, RSSI %i %s "
```
This print takes two full lines, when displaying this print on my terminal with the default terminal size configuration.
That makes it harder to to see what is happing. 

For me it would have been sufficient to only print address and device name.

cvinayak replied:
This is taken from the shell application.

rugeGerritsen commented on samples/bluetooth/per_sync/src/main.c:
```diff
@@ -0,0 +1,287 @@
+			return;
+		}
+		printk("Found periodic advertising.\n");
+
```
Did you consider stopping the scanner after the sync was established? 
I find it hard to see the prints related to the periodic advertiser in terminal output. It is overflowed by advertising reports from the extended advertiser.

cvinayak replied:
No, so as to be able to demo that periodic sync continues even when local events could be skipped due to scan events or possible extended advertising collisions on auxiliary channels from other devices.

But yes, if this is a sample to demo low power usage of periodic adv sync, scanning be disabled after sync established.

rugeGerritsen commented on samples/bluetooth/per_sync/src/main.c:
```diff
@@ -0,0 +1,287 @@
+	printk("success.\n");
+
+	printk("Start scanning...");
+	err = bt_le_scan_start(BT_LE_SCAN_ACTIVE, NULL);
```
Strictly speaking you don't need to perform active scanning when scanning for extended advertisers. I'm fine with keeping it as is.

cvinayak replied:
No action, as the sample will be able to demonstrate active extended scanning co-existing with periodic advertising synchronization. (yes, periodic advertising sync does not need active scanning)

rugeGerritsen commented on samples/bluetooth/per_sync/src/main.c:
```diff
@@ -0,0 +1,287 @@
+
+	bt_addr_le_to_str(info->addr, le_addr, sizeof(le_addr));
+
+	printk("PER_ADV_SYNC[%u]: [DEVICE]: %s, tx_power %i, "
```
To me the interesting parts is seeing that that the data is being updated. Therefore I would prioritize printing the data instead of printing the `tx_power` and `cte` field. Those are anyways set to "not present" for this sample

cvinayak replied:
As this sample establishes periodic adv sync to any periodic advertising in the vicinity, I will leave it to the user to parse the data as he desires, but yes, I will add display of the manufacturer data if present.

action: display of the manufacturer data.

cvinayak replied:
done

rugeGerritsen commented on samples/bluetooth/per_sync/src/main.c:
```diff
@@ -0,0 +1,287 @@
+		err = k_sem_take(&sem_per_sync, K_FOREVER);
+		if (err) {
+			printk("failed (err %d)\n", err);
+			continue;
```
I assume that we in all cases should return from main when we fail to take a semaphore with infinite timeout?
```suggestion
			return;
```

cvinayak replied:
action: I will use a timeout and cancel sync create. It will be difficult to test (to kill periodic advertiser before the scanner sync after it has found it).

cvinayak replied:
done

rugeGerritsen commented on samples/bluetooth/per_sync/src/main.c:
```diff
@@ -0,0 +1,287 @@
+	printk("Success.\n");
+
+	printk("Initialize semaphore sem_per_adv...");
+	err = k_sem_init(&sem_per_adv, 0, 1);
```
To simplify the sample, you could consider using static initialization of the semaphores using `K_SEM_DEFINE()`

cvinayak replied:
done

rugeGerritsen reviewed (commented):
I would prefer the samples to be named `periodic_adv` and `periodic_sync` to be more explicit. Users with only basic understanding of Bluetooth LE may not know if `per` is an abbreviation for `peripheral` or `periodic`.

cvinayak commented on include/bluetooth/hci.h:
```diff
@@ -1967,6 +1967,8 @@ struct bt_hci_evt_le_conn_complete {
 	uint8_t      clock_accuracy;
 } __packed;
 
+#define BT_HCI_LE_RSSI_NOT_AVAILABLE            0x7F
```
`RSSI is not available` with value 0x7F is described only for 4 Bluetooth LE events. All other events use RSSI value range. Correct me, if I have missed non-LE events that use the 0x7F explicit value, HCI_Read_RSSI uses 0x81 to indicate signal strength indication is not available.

rugeGerritsen replied:
HCI_Read_RSSI: "If the RSSI cannot be read the RSSI metric shall be set to 127 (0x7F)"

cvinayak replied:
Vol 5: AMP Host Controller Interface, Section 2.6, `The value shall be 0x81 (-127 dBm) if the signal strength indication is not
available.` Hence, retaining `BT_HCI_LE_RSSI_NOT_AVAILABLE` to specify LE HCI event value.

cvinayak commented on samples/bluetooth/per_adv/boards/nrf52840dk_nrf52840.conf:
```diff
@@ -0,0 +1,2 @@
+CONFIG_BT_CTLR_ADV_EXT=y
```
Not all boards use the native split controller.

cvinayak commented on samples/bluetooth/per_adv/src/main.c:
```diff
@@ -0,0 +1,71 @@
+
+#include <bluetooth/bluetooth.h>
+
+static uint8_t mfg_data[] = { 0xff, 0xff, 0x00 };
```
Manufacturer data, I just used this from the `scan_adv` sample.

rugeGerritsen replied:
What is the `g` abbreviating? Maybe naming it something as simple as `ad_data` is better?

cvinayak commented on samples/bluetooth/per_adv/src/main.c:
```diff
@@ -0,0 +1,71 @@
+static uint8_t mfg_data[] = { 0xff, 0xff, 0x00 };
+
+static const struct bt_data ad[] = {
+	BT_DATA(BT_DATA_MANUFACTURER_DATA, mfg_data, 3),
```
The `const` applies to the `ad` memory, `mfg_data` memory is non-const.

cvinayak commented on samples/bluetooth/per_sync/src/main.c:
```diff
@@ -0,0 +1,287 @@
+
+#define NAME_LEN    30
+
+static bool         is_per_adv;
```
`is_per_adv`: There `is` a `per`iodic `adv`ertising.

Thalley replied:
I guess it could be renamed into `per_adv_found` if that helps :)

cvinayak replied:
done

carlescufi commented:
@joerchan @jhedberg @rugeGerritsen could you review?

Moediddy901 commented:
I need help
